Zend/zend_inheritance.c

```diff
@@ -173,72 +174,205 @@ static zend_always_inline zend_bool zend_iterable_compatibility_check(zend_arg_i
 }
 /* }}} */
 
-static int zend_do_perform_type_hint_check(const zend_function *fe, zend_arg_info *fe_arg_info, const zend_function *proto, zend_arg_info *proto_arg_info) /* {{{ */
+/* todo: this is probably useful elsewhere too */
+/* caller is responsible for adding any necessary refcount */
+static zend_string *_get_parent_class_name(zend_class_entry *ce)
+{ /* {{{ */
+	zend_string *pname;
+
+	if (ce->ce_flags & ZEND_ACC_LINKED) {
+		ZEND_ASSERT(ce->parent);
+		pname = ce->parent->name;
+	} else {
+		pname = ce->parent_name;
+	}
+
+	/* If there is a parent, it must have a name */
+	ZEND_ASSERT(pname);
```

Comment:
This assertion will trigger for the following code:

```php
<?php
  
trait T {
    function foo(parent $x) {}
}

class A {
    use T;
}
class B extends A {
    function foo(parent $x) {}
}
```

Reply:
I think that should be fixed upstream -- that should be an error, since `A` has no parent. Is it not possible upstream? If not, some direction would be great so I can fix it correctly.

Reply:
We don't know whether or not `parent` is valid until the trait is used -- are you suggesting that we add a check for this as part of trait import?

Reply:
Yes, I think so. Maybe during compilation of the trait we can add a flag that indicates that it uses `parent`, and then on import we simply check for the flag. What do you think?